### ISSUE TITLE
Decryption works in build.rs but not when calling function from tests.

### ISSUE THREAD

**Issue by Nordgaren** [rust-lang/rust#108892]
```rust
        let mut aes_strings = vec![];
        let mut aes_hashes = vec![];
        for string_name in AES_STRINGS {
            let encrypted = aes_encrypt_bytes(string_name.as_bytes(), &aes_key, &aes_iv);
            aes_strings.push(AESString {
                string_name,
                encrypted,
                offset: usize::MAX,
            });

            //also store the hash, if it's needed.
            let hash = if string_name.ends_with(".dll") || string_name.ends_with(".DLL") {    
                hash_case_insensitive(string_name.as_ptr() as usize, string_name.len())
            } else {
                let mut c_string = string_name.to_string();
                c_string.push(0 as char);
                hash(string_name.as_ptr() as usize)
            };

            let encrypted = aes_encrypt_bytes(&hash.to_ne_bytes(), &aes_key, &aes_iv);
            //if I add these checks to each aes encrypted hash, it decrypts correctly
            if string_name == "NtFlushInstructionCache" {
            }
            if string_name == "GetProcAddress" {
            }
            if string_name == "VirtualAlloc" {
            }

            aes_hashes.push(AESHash {
                string_name,
                encrypted,
                offset: usize::MAX,
            })
        }
```

The above code is part of my `build.rs` script that generates a PE resource and embeds it into the main application. Currently working on a red team tool that can search for functions via hash or via string name, but for some reason when I aes 256 encrypt the hashes, they don't come out correctly on the other end, inside the actual program.

I have compared the key iv and encrypted bytes to the ones a runtime,and they are the correct bytes. I even had cyberchef decrypt them with the key and iv being output from `build.rs` and the encrypted data from the main program, and it gives me the correct value, as well.

After toying around a bit, I found these things out:

-  If I decrypt them in the build script, they decrypt properly.  
-  If I add a check for each string, like I do the above example, it "fixes" the issue.   
-  If I exchange un-commented line below for the ones in comments, it breaks the "fix" when checking each string (and I need to do the u16 encoding).    

```rust
               let hash = if string_name.ends_with(".dll") || string_name.ends_with(".DLL") {
               //let wString = string_name.encode_utf16().collect::<Vec<u16>>();
               //hash_case_insensitive(wString.as_ptr() as usize, string_name.len() * 2)
           
                hash_case_insensitive(string_name.as_ptr() as usize, string_name.len())
            } else {
                let mut c_string = string_name.to_string();
                c_string.push(0 as char);
                hash(string_name.as_ptr() as usize)
            };
```

Does anyone have a clue what is going on, here?

**Comment by Skgland** [comment]
I think the following could be a problem
```diff
- hash_case_insensitive(wString.as_ptr() as usize, string_name.len() * 2)
+ hash_case_insensitive(wString.as_ptr() as usize, wString.len() * 2)
```
The utf-16 string might not be twice as large as the utf-8 string.

**Comment by Nordgaren** [comment]
> I think the following could be a problem
> 
> ```diff
> - hash_case_insensitive(wString.as_ptr() as usize, string_name.len() * 2)
> + hash_case_insensitive(wString.as_ptr() as usize, wString.len() * 2)
> ```
> 
> The utf-16 string might not be twice as large as the utf-8 string.

Thanks for that! I'll make that change, so as not to cause problems in the future. 
This doesn't effect the issue, though. Both strings that are hashes as wide strings and regular C strings fail to decrypt :(

**Comment by Nordgaren** [comment]
This build script builds a `consts.rs` file for the location and length of all of the encrypted strings. I tried decrypting them by hand by printing them out and putting them in CyberChef. I get the same, incorrect, result from decrypting them there.

Is it possible that the wrong offset is being written to the consts file? I would think that random data would not decrypt properly.

It's just so weird that adding an if statement to check the string is what fixes the problem. Sorry for the bump, but this is some new info that I just discovered.
